// @linkedin/pemberly-core/lib/helpers/get-product-info.js
// Information about this current MP
const path = require('path');
const moment = require('dayjs');

// load product information in product-spec.json
module.exports = mpLocation => {
  try {
    const spec = require(path.join(mpLocation, 'product-spec.json'));
    const mpName = spec.name; // will be used to construct other fields
    return {
      mpName,
      org: spec.group ? spec.group.replace(/@.*@/, () => mpName) : mpName,
      version: spec.version,
      target: `${mpName}_trunk`,
      artifactName: `linkedin-${mpName}`,
      publication: moment().format('YYYYMMDDHHmmss'), // timestamp
      status: 'integration', // TODO: figure out intergration vs release
      artifact_type: 'lib', // TODO: what the heck is this
      ext: 'tgz',
      conf: 'archives,default', // TODO: what the hack is this
      root: mpLocation, // project root
      buildDir: path.join(mpLocation, 'build'),
      pemberlyBuildDir: path.join(mpLocation, 'build', 'pemberly-core'),
      publishDir: path.join(mpLocation, 'build', 'pemberly-core', 'publish'),
      versions: spec.build.versions, // dependencies version
      metaFile: path.join(mpLocation, 'build', 'pemberly-core', '.meta.json'), // to store info
      PACKAGE_JSON: path.join(mpLocation, 'package.json'),
    };
  } catch (e) {
    // if product-spec.json does not exist
    return {};
  }
};
